tools/autograd/load_derivatives.py

```diff
@@ -251,91 +250,88 @@ def stride_expr(name: str) -> str:
         (r'to_args_scalartypes\({}\)', {
             'suffix': '_args_scalartypes',
-            'type': 'std::vector<at::ScalarType>',
+            'nctype': lambda name: NamedCType(name, VectorCType(BaseCType(scalarTypeT))),
         }),
         # replace TensorGeometry(self) with self_geometry
         (r'TensorGeometry\({}\)', {
             'suffix': '_geometry',
-            'type': 'at::TensorGeometry',
+            'nctype': lambda name: NamedCType(name, BaseCType(tensorGeometryT)),
         }),
         (r'{}.scalar_type\(\)', {
             'suffix': '_scalar_type',
-            'type': 'at::ScalarType',
+            'nctype': lambda name: NamedCType(name, BaseCType(scalarTypeT)),
         }),
         # replace self.dim() with self_dim
         (r'{}.dim\(\)', {
             'suffix': '_dim',
-            'type': 'int64_t',
+            'nctype': lambda name: NamedCType(name, BaseCType(intT)),
         }),
         # replace self.strides() with self_strides
         (r'{}.strides\(\)', {
             'suffix': '_strides',
-            'type': 'at::IntArrayRef',
+            'nctype': lambda name: NamedCType(name, BaseCType(intArrayRefT)),
             'expr': stride_expr,
         }),
     ]
 
     # find which arguments need to be saved
     saved: List[SavedAttribute] = []
 
-    for name, type in zip(arg_names, arg_types):
+    for nctype in nctypes:
         # First search the formula for expressions which can be evaluated
         # when the autograd Function is created to avoid saving variables
         for regex, info in REPLACEMENTS:
             def repl(m: Match[str]) -> str:
                 suffix: str = info['suffix'](m) if callable(info['suffix']) else info['suffix']
-                expr: str = info['expr'](name) if 'expr' in info else m.group(0)
+                expr: str = info['expr'](nctype.name) if 'expr' in info else m.group(0)
                 saved.append(SavedAttribute(
-                    name=name + suffix,
-                    type=info['type'],
+                    nctype=info['nctype'](nctype.name + suffix),
                     expr=expr,
                 ))
                 if 'res' in info:
-                    replacement: str = info['res'](name)
+                    replacement: str = info['res'](nctype.name)
                     return replacement
-                return name + suffix
+                return nctype.name + suffix
 
-            formula = re.sub(regex.format(name), repl, formula)
+            formula = re.sub(regex.format(nctype.name), repl, formula)
 
         # Find any variables which remain in the formula and save them
-        if re.search(IDENT_REGEX.format(name), formula):
+        if re.search(IDENT_REGEX.format(nctype.name), formula):
             saved.append(SavedAttribute(
-                name=name,
-                # TODO: change from string to type data model
-                type=type.replace('const ', '').replace(' &', ''),
```

Comment:
I added a `remove_const_ref()` method to `CType` to emulate this in the data model.... it feels a little better, but it also required copying it into the 9 different `CType` classes we now have.